[Data Frame] Avoid assertions on empty Optional in DF usage test

davidkyle opened the pull request:
Avoid assertions on an empty Optional in DF usage test where randomisation means a reduce operation may occur on a zero lenght list resulting in an empty Optional. 

Closes #40022

davidkyle committed:
Avoid assertions on empty list

elasticmachine commented:
Pinging @elastic/ml-core

davidkyle committed:
fix checkstyle

hendrikmuhs commented on x-pack/plugin/data-frame/src/test/java/org/elasticsearch/xpack/dataframe/DataFrameFeatureSetTests.java:
```diff
@@ -142,16 +142,19 @@ public void testUsage() throws InterruptedException, ExecutionException, IOExcep
+                Optional<DataFrameIndexerTransformStats> combinedStats = transformsStateAndStats.stream().map(x -> x.getTransformStats())
+                        .reduce((l, r) -> l.merge(r));
+
+                if (combinedStats.isPresent()) {
```
I think the problem of this test is on line 131, the `if` is broken, it broke because `transformConfigWithoutTasks` is empty but `transformsStateAndStats` isn't. 

I admit this isn't a master piece in test writing (I think I originally wrote it), it would be clearer and cleaner if we split this test into the empty and non-empty case. We then do not need to workaround here, which I am not sure about: it disables all asserts and tests nothing if `isPresent()` returns false, which means we have no coverage for this case.

benwtrent replied:
@hendrikmuhs looking at the test we do have assertions for when `combinedStats.isPresent()`, they all occur earlier in the test for asserting the number of DFs that are given in certain states. I am not sure what `else` tests would exist in this case. 

I should have caught this in my refactor,

davidkyle replied:
There are assertions on `transformConfigWithoutTasks` lines 143-144 even if `transformsStateAndStats` is empty. 

I pushed a small change so it asserts the stats have value 0 if `transformsStateAndStats` is empty

davidkyle committed:
Refactor to test case where no stats

davidkyle committed:
Factor out dataframe get usage logic

davidkyle commented:
I refactored `DataFrameFeatureSetUsage.usage` to extract the logic into a separate method. That makes the testing easier as it removes the need to mock client responses. Then I ran the test 100 times and hit failures because the random task Ids were not unique so I reworked the test to prevent that happening.

benwtrent reviewed (approved):
👍 Integration with the usage as a whole is already done in `org.elasticsearch.xpack.dataframe.integration.DataFrameUsageIT`. So, pulling out the logic and testing it outside of using the client is cool

hendrikmuhs reviewed (approved):
LGTM
